src/runtime_src/core/pcie/windows/device_windows.cpp

```diff
@@ -911,6 +919,65 @@ struct data_retention
 
 };
 
+struct mailbox
+{
+	using result_type = boost::any;
+
+	static xcl_mailbox
+		init_mailbox_info(const xrt_core::device* dev)
+	{
+		xcl_mailbox info = { 0 };
+		userpf::get_mailbox_info(dev->get_user_handle(), &info);
+		return info;
+	}
+
+	static result_type
+		get_info(const xrt_core::device* device, key_type key)
+	{
+		static std::map<const xrt_core::device*, xcl_mailbox> info_map;
+		static std::mutex mutex;
+		std::lock_guard<std::mutex> lk(mutex);
+
+		auto it = info_map.find(device);
+		if (it == info_map.end()) {
+			auto ret = info_map.emplace(device, init_mailbox_info(device));
+			it = ret.first;
+		}
+
+		const xcl_mailbox& info = (*it).second;
+		int count = 0, i, size;
+		char *buf = new char[1024];
+		std::vector<std::string> vec;
+		switch (key) {
+		case key_type::mailbox_metrics:
+			count += sprintf(buf, "raw bytes received: %d\n", info.mbx_recv_raw_bytes);
+			vec.push_back(std::string(buf, buf + count));
+			for (i = 0; i < XCL_MAILBOX_REQ_MAX; i++) {
+				size = sprintf(buf + count, "req[%d] received: %d\n", i, info.mbx_recv_req[i]);
+				vec.push_back(std::string(buf + count, buf + count + size));
+				count += size;
+			}
```

Comment:
This code should be re-written using c++ standards. 
- add a comment of what kind of data the IOCTL is returning, i.e., what are we trying to parse
- sprintf is a C function, so please use boost::format instead
- It is safer to use `std::vector<char> buf(1024);` than `char *buf = new char[1024];`, but I am not sure if you'll need it after using boost::format